sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/CatalogManager.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalog.v2
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * A thread-safe manager for [[CatalogPlugin]]s. It tracks all the registered catalogs, and allow
+ * the caller to look up a catalog by name.
+ */
+class CatalogManager(conf: SQLConf) extends Logging {
+
+  private val catalogs = mutable.HashMap.empty[String, CatalogPlugin]
+
+  def catalog(name: String): CatalogPlugin = synchronized {
+    catalogs.getOrElseUpdate(name, Catalogs.load(name, conf))
+  }
+
+  def defaultCatalog: Option[CatalogPlugin] = {
+    conf.defaultV2Catalog.flatMap { catalogName =>
+      try {
+        Some(catalog(catalogName))
+      } catch {
+        case NonFatal(e) =>
+          logError(s"Cannot load default v2 catalog: $catalogName", e)
+          None
+      }
+    }
+  }
+
+  def v2SessionCatalog: Option[CatalogPlugin] = {
+    try {
+      Some(catalog(CatalogManager.SESSION_CATALOG_NAME))
+    } catch {
+      case NonFatal(e) =>
+        logError("Cannot load v2 session catalog", e)
+        None
+    }
+  }
+
+  private def getDefaultNamespace(c: CatalogPlugin) = c match {
+    case c: SupportsNamespaces => c.defaultNamespace()
+    case _ => Array.empty[String]
+  }
+
+  private var _currentNamespace = {
+    // The builtin catalog use "default" as the default database.
```

Comment:
I'm thinking more about whether this should be the duty of the CatalogManager or the catalog itself...

My thoughts are:
  1. Once the V2SessionCatalog `SupportsNamespaces`, it can have the default namespace `default` set within it
  2. When you run a command like `USE foo.bar` where `foo` is the catalog, we should set the default namespace within that catalog to be `bar`.
  3. What if the `default` catalog doesn't support namespaces in the first place?

I'm a bit worried that having the default namespace setting in the CatalogManager can lead to a split brain situation.

Reply:
I think current namespace only make sense to the current catalog, e.g. `SELECT ... FROM t`, `t` can be a table in the current catalog's current namespace. However, for `SELECT ... FROM c1.t`, it's confusing to say `t` is a table in catalog `c1`'s current namespace.

When a table identifier starts with a catalog name, it should be a fully qualified identifier, and we can't apply current namespace here.

catalog (including `V2SessionCatalog`) can report its default namespace, which will be used as the current namespace when switching to the catalog at the first time.

Reply:
I think we're saying the same things. I totally agree that:
for a catalog `c1` ```sql
SELECT ... FROM c1.t
```, this should be a fully qualified identifier. I'm saying that we should push the namespace configuration into the catalog that supports it. It shouldn't be part of the CatalogManager.

Reply:
I'm not sure we should ask the catalog implementation to do it. This will make catalog mutable and make it harder to implement. e.g. they need to take care of the current namespace in many methods like `loadTable`, `createTable`, etc. It also brings some risks if the catalog implementation has bugs for tracking the current namespace.

Instead, I think it's better to let Spark track the current namespace, and always extend the table identifier with the current namespace and make it a fully qualified identifier before calling catalog APIs. I'm fine for not tracking it in `CatalogManager`, but we should track it somewhere in Spark.

Reply:
A side story: Hive can track the current database, but Spark track the current database itself, and always call Hive APIs with fully qualified table identifier.

Reply:
Oh, wait, I got the defaultNamespace setting in `SupportsNamespaces` wrong. I thought it was actually tracked there. This looks good to me. Sorry.